### PR TITLE
port: teams sso token exchange middleware

### LINKED ISSUE #3568: port: Add Teams SSO Token Exchange middleware (#5475)
The changes in [Add Teams SSO Token Exchange middleware (#5475)](https://github.com/microsoft/botbuilder-dotnet/pull/5475) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
Fixes https://github.com/microsoft/botbuilder-dotnet/issues/5473
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by joshgummersall**
Fixes #3568

### CODE REVIEW

**Review by EricDahlvang — APPROVED** [review]
tested, confirmed it works with memory storage and cosmosdb

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    const activity = context.activity;
+
+    const channelId = activity.channelId;
+    if (!channelId) {
+        throw new Error('invalid activity-missing channelId');
+    }
+
+    const conversationId = activity.conversation?.id;
+    if (!conversationId) {
+        throw new Error('invalid activity-missing Conversation.Id');
```
Comment (stevengum) [inline libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts:28]: nit: Inconsistent error message formatting. Can you align on the format used and clean up the casing on references to members?

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    TokenResponse,
+} from 'botframework-schema';
+
+function getStorageKey(context: TurnContext): string {
+    const activity = context.activity;
+
+    const channelId = activity.channelId;
+    if (!channelId) {
+        throw new Error('invalid activity-missing channelId');
+    }
```
Comment (stevengum) [inline libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts:24]: In the `TeamsSSOTokenExchangeMiddleware`, does the value of the channelId need to be [`botframework-schema.Channels.Msteams`](https://github.com/microsoft/botbuilder-js/blob/ada50e897946277c46474911f0cd6bbd42fe8e3b/libraries/botframework-schema/src/index.ts#L2116) or should this middleware fire for any signin/tokenExchange?

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+/**
+ * If the activity name is signin/tokenExchange, this middleware will attempt to
+ * exchange the token, and deduplicate the incoming call, ensuring only one
+ * exchange request is processed.
+ *
+ * If a user is signed into multiple Teams clients, the Bot could receive a
+ * "signin/tokenExchange" from each client. Each token exchange request for a
+ * specific user login will have an identical Activity.Value.Id.
```
Comment (stevengum) [inline libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts:46]: nit: member casing.

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+ *
+ * Only one of these token exchange requests should be processed by the bot.
+ * The others return <see cref="System.Net.HttpStatusCode.PreconditionFailed"/>.
```
Comment (stevengum) [inline libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts:49]: Incorrect language.

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+ * For a distributed bot in production, this requires a distributed storage
+ * ensuring only one token exchange is processed. This middleware supports
+ * CosmosDb storage found in Microsoft.Bot.Builder.Azure, or MemoryStorage for
```
Comment (stevengum) [inline libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts:52]: Should be `botbuilder-azure`

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+ * local development. IStorage's ETag implementation for token exchange activity
```
Comment (stevengum) [inline libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts:53]: nit: `Storage` not `IStorage`

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * Called each time the bot receives a new request.
+     *
+     * @param context Context for current turn of conversation with the user.
+     * @param next Function to call to continue execution to the next step in the middleware chain.
+     */
+    async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
+        if (context.activity.name === signInConstants.tokenExchangeOperationName) {
+            // If the TokenExchange is NOT successful, the response will have already been sent by ExchangedTokenAsync
```
Comment (stevengum) [inline libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts:73]: Should be `exchangedToken`, not `ExchangedTokenAsync`

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+            if (!(await this.exchangedToken(context))) {
+                return;
+            }
+
+            // Only one token exchange should proceed from here. Deduplication is performed second because in the case
+            // of failure due to consent required, every caller needs to receive the
```
Comment (stevengum) [inline libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts]: nit: incomplete comment

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    private async exchangedToken(context: TurnContext): Promise<boolean> {
+        let tokenExchangeResponse: TokenResponse;
+        const tokenExchangeRequest: TokenExchangeInvokeRequest = context.activity.value;
+
+        const tokenProvider = (context.adapter as unknown) as ExtendedUserTokenProvider;
+        if (!tokenProvider) {
+            throw new Error('Token Exchange is not supported by the current adapter.');
+        }
```
Comment (stevengum) [inline libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:134]: Shouldn't this be an assertion? At runtime `context.adapter` is always true unless `context.adapter` was changed to a falsy value.

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts (hunk 2 above)
Comment (joshgummersall) [reply libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts]: I ported this closely from .NET - @EricDahlvang ?

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    TokenResponse,
+} from 'botframework-schema';
+
+function getStorageKey(context: TurnContext): string {
+    const activity = context.activity;
+
+    const channelId = activity.channelId;
+    if (!channelId) {
+        throw new Error('invalid activity-missing channelId');
+    }
```
Comment (stevengum) [inline libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:24]: The `TeamsSSOTokenExchangeMiddleware` implicitly supports `signin/tokenExchange` Activities for any channel?

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+    const conversationId = activity.conversation?.id;
+    if (!conversationId) {
+        throw new Error('invalid activity-missing Conversation.Id');
```
Comment (stevengum) [inline libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:28]: Inconsistent Error message formatting and the members are incorrectly cased.

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+ * If the activity name is signin/tokenExchange, this middleware will attempt to
+ * exchange the token, and deduplicate the incoming call, ensuring only one
+ * exchange request is processed.
+ *
+ * If a user is signed into multiple Teams clients, the Bot could receive a
+ * "signin/tokenExchange" from each client. Each token exchange request for a
+ * specific user login will have an identical Activity.Value.Id.
+ *
+ * Only one of these token exchange requests should be processed by the bot.
+ * The others return <see cref="System.Net.HttpStatusCode.PreconditionFailed"/>.
```
Comment (stevengum) [inline libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:49]: Reference to .NET code.

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+ * For a distributed bot in production, this requires a distributed storage
+ * ensuring only one token exchange is processed. This middleware supports
+ * CosmosDb storage found in Microsoft.Bot.Builder.Azure, or MemoryStorage for
+ * local development. IStorage's ETag implementation for token exchange activity
```
Comment (stevengum) [inline libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:53]: Should be `botbuilder-azure` and `Storage`

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+
+    /**
+     * Called each time the bot receives a new request.
+     *
+     * @param context Context for current turn of conversation with the user.
+     * @param next Function to call to continue execution to the next step in the middleware chain.
+     */
+    async onTurn(context: TurnContext, next: () => Promise<void>): Promise<void> {
+        if (context.activity.name === signInConstants.tokenExchangeOperationName) {
+            // If the TokenExchange is NOT successful, the response will have already been sent by ExchangedTokenAsync
```
Comment (stevengum) [inline libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:73]: Should be `exchangedToken`

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+            if (!(await this.exchangedToken(context))) {
+                return;
+            }
+
+            // Only one token exchange should proceed from here. Deduplication is performed second because in the case
+            // of failure due to consent required, every caller needs to receive the
```
Comment (stevengum) [inline libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:79]: Incomplete comment

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts (hunk 8 above)
Comment (joshgummersall) [reply libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts]: They needed you to review the .NET PR!

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+                return false;
+            }
+
+            throw err;
+        }
+
+        return true;
+    }
+
+    private async sendInvokeResponse(
```
Comment (stevengum) [inline libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:116]: Could be static or internal (not PR blocking)

Comment (joshgummersall) [reply libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:116]: Good suggestion!

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts (hunk 10 above)
Comment (joshgummersall) [reply libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts]: cc @EricDahlvang

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts (hunk 9 above)
Comment (joshgummersall) [reply libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts]: Rewrote using stdlib assertions.

File: libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts (hunk 2 above)
Comment (joshgummersall) [reply libraries/botbuilder-core/src/teams/teamsSSOTokenExchangeMiddleware.ts]: Deferring to the other comment for discussion.

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts (hunk 10 above)
Comment (EricDahlvang) [reply libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts]: As far as I know, these only come from Teams? 

Other channels wouldn't need this middleware anyway.  The true purpose of the middleware is to ensure duplicate token exchange requests are not processed (only one can be processed).

Comment (joshgummersall) [reply libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts]: Theoretically, then, we could add a filter specifically for events from the teams channel?

Comment (joshgummersall) [reply libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts]: Bookmarking with https://github.com/microsoft/botbuilder-js/issues/3574 for future consideration.

File: libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+    await context.sendActivity({
+        type: ActivityTypes.InvokeResponse,
+        value: { body, status },
+    });
+}
+
+const assertExchangeToken: Assertion<Pick<ExtendedUserTokenProvider, 'exchangeToken'>> = (val, path) => {
+    assert.unsafe.castObjectAs<ExtendedUserTokenProvider>(val, path);
+    assert.func(val.exchangeToken, path.concat('exchangeToken'));
+};
```
Comment (stevengum) [inline libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts:55]: Thanks! I know we could use this assertion in the OAuthPrompt and probably OAuthInput.

**Review by stevengum — APPROVED** [review]
:shipit:

Like @johntaylor said, let's get some tests [added](https://github.com/microsoft/botbuilder-dotnet/pull/5475#pullrequestreview-635075439)

**Pull request merged**